### PR TITLE
Set Title and start date and time for duplicate events using duplicate WFH

### PR DESCRIPTION
**Pull request by tyfranck**
For rolling forward Lectures from one year to the next we need to be able to change the Title and Start Date and Time of the duplicated mediapackage.

This PR is based on PR #3409 .

Fix 2 issues highlighted in the comments of the above PR
- Date changes not reflected
- issues with empty date

### Your pull request should…

* [x] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [x] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [x] include migration scripts and documentation, if appropriate
* [x] pass automated tests
* [x] have a clean commit history
* [x] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

### CODE REVIEW

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java [hunk 1]
```diff
@@ -227,7 +246,7 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
 
     SeriesInformation series = null;
     AccessControlList seriesAccessControl = null;
-    if (!seriesId.isEmpty()) {
+    if (!seriesId.isEmpty() && !seriesId.startsWith("${")) {
```
Comment (gregorydlogan) [inline modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: This doesn't check that the substitution terminates.  Is that expected?

**Review by gregorydlogan — APPROVED** [review]
LGTM, aside from my comment.

File: docs/guides/admin/docs/workflowoperationhandlers/duplicate-event-woh.md [hunk 2]
```diff
@@ -16,8 +16,8 @@ For each duplicated event the new media package ID is stored as a workflow prope
 
 |Name                                    |Example                                                             |Description                                    |
 |----------------------------------------|--------------------------------------------------------------------|-----------------------------------------------|
-|duplicate\_media\_package\_*number*\_id |`duplicate_media_package_1_id=e72f2265-472a-49ae-bc04-8301d94b4b1a` |Media package ID of the duplicated event       | \*
-|duplicate\_media\_package\_ids          |`duplicate_media_package_ids=e72f2265-472a-49ae-bc04-8301d94b4b1a, a32e2265-472a-49ae-bc04-8351d94b4b1c` | comma separated list of Media package IDs of the duplicated event |
+|duplicate_media_package_*number*_id     |`duplicate_media_package_1_id=e72f2265-472a-49ae-bc04-8301d94b4b1a` |Media package ID of the duplicated event       | \*
```
Comment (lkiesow) [inline docs/guides/admin/docs/workflowoperationhandlers/duplicate-event-woh.md:19]: Why did you remove the escape character on the underscore character?

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java (hunk 1 above)
Comment (lkiesow) [inline modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: The `.startsWith("${")` shouldn't be in here. If that's a value, it should be used

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java [hunk 3]
```diff
@@ -309,14 +328,45 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
...
+      final Date startDate;
       try {
         String newMpId = workflowInstance.getConfiguration("newMpId");
         if (newMpId == null) {
           newMpId = UUID.randomUUID().toString();
         }
         // Clone the media package (without its elements)
-        newMp = copyMediaPackage(mediaPackage, series, newMpId, noSuffix, i + 1, copyNumberPrefix);
+        String useTitle;
+        if (title.isEmpty() || (title.startsWith("${") && title.endsWith("}"))) {
```
Comment (lkiesow) [inline modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: Same here, the `(title.startsWith("${") && title.endsWith("}"))` is weird

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java [hunk 4]
```diff
@@ -309,14 +328,45 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
...
+          useTitle = mediaPackage.getTitle();
```
Comment (lkiesow) [inline modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: This method is deprecated. It's not guaranteed to yield the correct title. Get metadata from the DC catalog

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java [hunk 5]
```diff
@@ -309,14 +328,45 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
...
+        }
+        final String newTitle = noSuffix
+            ? useTitle
+            : String.format("%s (%s %d)", useTitle, copyNumberPrefix, i + 1);
+        Date mpDate = mediaPackage.getDate();
+        String date = new SimpleDateFormat("yyyy-MM-dd").format(mpDate);
+        String time = new SimpleDateFormat("HH:mm:ss").format(mpDate);
+        if (!startDateString.isEmpty()) {
+          String [] dt = startDateString.trim().split("\\s+");
+          if (dt.length >= 2) {
```
Comment (lkiesow) [inline modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: If `length > 2` that's clearly an error and you shouldn't guess values but throw an error

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java [hunk 6]
```diff
@@ -309,14 +328,45 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
...
+            if (!dt[0].startsWith("${")) {
+              date = dt[0];
+            }
+            if (!dt[1].startsWith("${")) {
+              time = (dt[1].split(":").length == 2) ? dt[1] + ":00" : dt[1];
```
Comment (lkiesow) [inline modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: The documentation above states `HH:MM:SS` as the format, this allows `HH:MM:SS` and `HH:MM`.

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java [hunk 7]
```diff
@@ -309,14 +328,45 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
...
+            }
+          }
+        }
+        try {
+          mpDate = ADMIN_UI_DATE_FORMAT.parse(date + " " + time);
+        } catch (ParseException ex) {
+          logger.info("{} could not be parsed as Date", date + " " + time);
```
Comment (lkiesow) [inline modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: This should fail the operation instead of doing something unexpected

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java [hunk 8]
```diff
@@ -309,14 +328,45 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
...
+        }
+        logger.info("setting StartDate to {}", mpDate.toString());
```
Comment (lkiesow) [inline modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]:
```suggestion
        logger.info("setting StartDate to {}", mpDate);
```

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java [hunk 9]
```diff
@@ -309,14 +328,45 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
...
+        startDate = mpDate;
```
Comment (lkiesow) [inline modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: Do we need both variables?

File: etc/workflows/duplicate-event.xml [hunk 10]
```diff
@@ -12,18 +12,48 @@
...
+          </div>
+          <br/>
+          <div>
+            <label for="seriesId" style="width: 150px;" >Series Id:</label>
+            <input id="seriesId" name="seriesId" type="text" maxlength="255" style="width: 400px;" value=" " class="configField" />
+          </div>
+          <br/>
+          <div>
+            <label for="startDate" style="width: 150px;">StartDate (YYYY-MM-DD):</label>
+            <input id="startDate" name="startDate" type="text" style="width: 200px;" value=" " class="configField" />
```
Comment (lkiesow) [inline etc/workflows/duplicate-event.xml]: Use [`<input type="date">`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date)

File: etc/workflows/duplicate-event.xml [hunk 11]
```diff
@@ -12,18 +12,48 @@
...
+          </div>
+          <div>
+            <label for="startTime" style="width: 150px;">startTime (HH:MM:SS):</label>
+            <input id="startTime" name="startTime" type="text" style="width: 200px;" value=" " class="configField" />
```
Comment (lkiesow) [inline etc/workflows/duplicate-event.xml]: …and [`type=time`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/time)

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java [hunk 12]
```diff
@@ -309,14 +328,45 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, fi
...
-        newMp = copyMediaPackage(mediaPackage, series, newMpId, noSuffix, i + 1, copyNumberPrefix);
+        String useTitle;
+        if (title.isEmpty() || (title.startsWith("${") && title.endsWith("}"))) {
+          useTitle = mediaPackage.getTitle();
+        } else {
+          useTitle = title;
+        }
+        final String newTitle = noSuffix
+            ? useTitle
+            : String.format("%s (%s %d)", useTitle, copyNumberPrefix, i + 1);
```
Comment (lkiesow) [inline modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: I did not go through all the code but this seems a complex way of:

```suggestion
        if (title.isEmpty()) {
          title = mediaPackage.getTitle();
        }
        if (!noSuffix) {
          title = String.format("%s (%s %d)", title, copyNumberPrefix, i + 1);
        }
```

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java [hunk 13]
```diff
@@ -516,11 +565,16 @@ private MediaPackage copyDublinCore(
...
       final List<String> addTags,
       final List<String> overrideTags,
-      final List<URI> temporaryFiles) throws WorkflowOperationException {
+      final List<URI> temporaryFiles,
+      final Date creationDate
+  ) throws WorkflowOperationException {
     final DublinCoreCatalog destinationDublinCore = DublinCoreUtil.loadEpisodeDublinCore(workspace, source).get();
     destinationDublinCore.setIdentifier(null);
     destinationDublinCore.setURI(sourceDublinCore.getURI());
+    destinationDublinCore.set(DublinCore.PROPERTY_CREATED, OpencastMetadataCodec.encodeDate(creationDate,Precision.Second));
```
Comment (lkiesow) [inline modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]:
```suggestion
    destinationDublinCore.set(DublinCore.PROPERTY_CREATED, OpencastMetadataCodec.encodeDate(creationDate, Precision.Second));
```

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java [hunk 14]
```diff
@@ -516,11 +565,16 @@ private MediaPackage copyDublinCore(
...
     destinationDublinCore.set(DublinCore.PROPERTY_TITLE, destination.getTitle());
+    DublinCoreValue eventTime = EncodingSchemeUtils.encodePeriod(new DCMIPeriod(creationDate, creationDate), Precision.Second);
```
Comment (lkiesow) [inline modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: This essentially sets the event date to 0.
If we don't have a period, we shouldn't set it

**Review by lkiesow — CHANGES_REQUESTED** [review]

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java (hunk 1 above)
Comment (tyfranck) [reply modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: When a series is not provided by the user, `${seriesId}` in the workflow definition is not substituted. So without that check the series will literally be set as `${seriesId}` in such instance.

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java (hunk 3 above)
Comment (tyfranck) [reply modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: Same answer as above

File: docs/guides/admin/docs/workflowoperationhandlers/duplicate-event-woh.md (hunk 2 above)
Comment (tyfranck) [reply docs/guides/admin/docs/workflowoperationhandlers/duplicate-event-woh.md:19]: No idea why Tobias removed the escape character but looking at it closing, I can see that escape characters are not really needed in the name. There's no need to escape underscore character in the name.  So I guess it's a bit of clean up.

File: etc/workflows/duplicate-event.xml (hunk 10 above)
Comment (tyfranck) [reply etc/workflows/duplicate-event.xml]: Yeah, I used date as a type before but the trouble is you can't really set a default value as `value =""` (empty). 
So when the date is not set, a null value is sent for that property which in turn throws an exception.
Properties values are enforced in `org.opencastproject.assetmanager.impl.persistence` package by the following code:

```
  public Property toProperty() {
    final PropertyId id = new PropertyId(mediaPackageId, namespace, propertyName);
    if (stringValue != null) {
      return new Property(id, Value.mk(stringValue));
    } else if (dateValue != null) {
      return new Property(id, Value.mk(dateValue));
    } else if (longValue != null) {
      return new Property(id, Value.mk(longValue));
    } else if (boolValue != null) {
      return new Property(id, Value.mk(boolValue));
    } else {
      throw new RuntimeException("Bug. At least one of the value columns must be non null.");
    }
  }
```
Basically input of type date or time can't be used unless a default value is set. Or maybe you have some suggestions.

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java (hunk 4 above)
Comment (tyfranck) [reply modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: OK. That method needs @Deprecated Annotation if it can't been used anymore.

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java (hunk 1 above)
Comment (tyfranck) [reply modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: fixed

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java (hunk 5 above)
Comment (tyfranck) [reply modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: fixed

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java (hunk 6 above)
Comment (tyfranck) [reply modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: Documentation updated to mention both format

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java (hunk 7 above)
Comment (tyfranck) [reply modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: fixed

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java (hunk 9 above)
Comment (tyfranck) [reply modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: fixed

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java (hunk 12 above)
Comment (tyfranck) [reply modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: code simplified :-)

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java (hunk 14 above)
Comment (tyfranck) [reply modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: fixed

**Comment by tyfranck** [comment]
I've addressed most of the comments given my time. @lkiesow  I commented the one I didn't address. I don't see them as stopper but I let you judge.  Any suggestions welcome.

File: etc/workflows/duplicate-event.xml (hunk 10 above)
Comment (mliradelc) [reply etc/workflows/duplicate-event.xml]: In that case, where the time is null, I suggest using the current time and log a warning for the event.

File: etc/workflows/duplicate-event.xml (hunk 11 above)
Comment (mliradelc) [reply etc/workflows/duplicate-event.xml]: Same as above answer

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java (hunk 13 above)
Comment (mliradelc) [reply modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: Missing space here

**Comment by lkiesow** [comment]
@mliradelc if you are happy with the current state, feel free to merge.
We start the release process right now but still need some time to prepare the release notes.

**Review by mliradelc — CHANGES_REQUESTED** [review]
Hi @tyfranck I just reviewed your code, is good to merge.
 
There are some things that can be improved:
- Use the DC catalog instead of deprecated method
- When the time and date values are missing, use the actual time and date, then log a warning.
- And a missing space.

I added the comments that @lkiesow has pointed out.

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java (hunk 4 above)
Comment (tyfranck) [reply modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: fixed

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java (hunk 6 above)
Comment (tyfranck) [reply modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: fixed with code revamped

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java (hunk 8 above)
Comment (tyfranck) [reply modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: fixed

File: etc/workflows/duplicate-event.xml (hunk 10 above)
Comment (tyfranck) [reply etc/workflows/duplicate-event.xml]: fixed, current datetime set default in UI.

File: etc/workflows/duplicate-event.xml (hunk 11 above)
Comment (tyfranck) [reply etc/workflows/duplicate-event.xml]: fixed with code revamped.

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java (hunk 13 above)
Comment (tyfranck) [reply modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/DuplicateEventWorkflowOperationHandler.java]: fixed

**Comment by tyfranck** [comment]
Hi @mliradelc, 
Thanks for taking a look. With the latest update to this PR, I've addressed your points raised above.
1- Using DC to get mediapackage Title. 
2- Current date time is set as default in UI
3- Space removed.
4- Documentation updated accordingly to reflect new changes.

**Comment by tyfranck** [comment]
Any hint on why this is failing `Test configuration` run ?

File: etc/workflows/duplicate-event.xml [hunk 15]
```diff
@@ -35,6 +60,28 @@
...
+            return this;
+          }
+          $(this).val(formattedDateTime);
+          return this;
+        }
+
+        $(function () {
+            $('input[type="datetime-local"]').setNow();
+        });
+
```
Comment (mliradelc) [inline etc/workflows/duplicate-event.xml]: Here is the error, you need to delete the trailing spaces, that's why is failing the test

`etc/workflows/duplicate-event.xml:84`

**Review by mliradelc — CHANGES_REQUESTED** [review]

File: etc/workflows/duplicate-event.xml (hunk 15 above)
Comment (tyfranck) [reply etc/workflows/duplicate-event.xml]: Thanks. Updated now.

**Review by mliradelc — APPROVED** [review]
